cpp/src/arrow/engine/substrait/relation_internal.h

```diff
@@ -30,8 +30,17 @@
 namespace arrow {
 namespace engine {
 
+/// Information resulting from converting a Substrait relation.
+struct DeclarationInfo {
+  /// The compute declaration produced thus far.
+  compute::Declaration declaration;
+
+  /// The number of columns returned by the declaration.
+  size_t num_columns;
```

Comment:
Just curious, should this be just `int` or `size_t`. The `Table` and `RecordBatch` interfaces provide `num_columns` as int.

Reply:
Right, this can be an `int`.

Reply:
Seems weird to me to use a signed integer for a size; IMO `Table` and `RecordBatch` are wrong to do that (if the 32/64-bitness is an issue, at least use `uint32_t` or `unsigned int`). Now you'd always get warnings when comparing with the `size_t`'s returned by STL containers.

That said, I guess I won't argue with precedence...

Reply:
96f8c93

Reply:
For my curiosity: 

I understand your point. I am not an expert on the exact usage, although I assumed the current customs were adopted based on a certain logic. But for better understanding I looked into CUDF which require to expose a similar attribute as follows.

https://github.com/pytorch/pytorch/blob/5f8c2076dfd834b21cfafe2f35d19c3d02a958dc/caffe2/core/tensor.h#L386

Here they use size_type (I guess that is the same as size_t). 

PS: Numpy and PyTorch both represents the dimension in `int` or `int64_t` but I guess it can be negative according to some definitions (I am not entirely clear on the possibility).

cc @pitrou @wesm

Reply:
@jvanstraten Right, it's basically the convention used accross Arrow. That said, I do not really find it weird because frequently you'll act on sizes by adding/removing offsets from them - and those offsets are usually signed. It makes sense to make all integers signed, IMHO.